monai/losses/focal_loss.py

```diff
@@ -142,11 +140,12 @@ def forward(self, input: torch.Tensor, target: torch.Tensor) -> torch.Tensor:
             at = class_weight[None, :, None]  # N => 1,N,1
             at = at.expand((t.size(0), -1, t.size(2)))  # 1,N,1 => B,N,H*W
             # Multiply the log proba by their weights.
-            logpt = logpt * at
+            ce = ce * at
 
         # Compute the loss mini-batch.
-        weight = torch.pow(-pt + 1.0, self.gamma)
-        loss = torch.mean(-weight * t * logpt, dim=-1)
+        p = F.logsigmoid(-i * (t * 2.0 - 1.0))
+        loss = torch.mean((p * self.gamma).exp() * ce, dim=-1)
```

Comment:
Should this be `p**self.gamma`? Raising p to this power shows up in most other implementations of focal loss.

Reply:
This version is adapted from a codebase used by our kaggle participants, it's reliable according to their recommendations... This `p` is a `logsigmoid` instead of `sigmoid`, `exp(log_p * gamma) = (exp(log_p))**gamma`, I think it'll reduce the change of overflow for `exp(x)` when x is large

Reply:
OK, this implementation isn't like anything else I'm finding while looking for focalloss implementations in Pytorch.

Reply:
I had a quick search, a similar idea is presented here https://www.kaggle.com/rejpalcz/focalloss-for-keras

Reply:
Most exact implementation is [here](https://becominghuman.ai/investigating-focal-and-dice-loss-for-the-kaggle-2018-data-science-bowl-65fb9af4f36c) farther down the page.

Reply:
that's almost the same as the one in this PR, correct? https://gist.github.com/AdrienLE/bf31dfe94569319f6e47b2de8df13416#file-focal_dice_1-py

Reply:
It's the same gist actually. I think there's multiple sources that come back to the same place.